### PR TITLE
Trac 32417: Media widget

### PR DESCRIPTION
**Pull request by westonruter**
https://core.trac.wordpress.org/ticket/32417

### CODE REVIEW

**Comment by westonruter** [comment]
Issues reported by linters:

## JSCS

* [ ] src/wp-includes/js/wp-media-widget.js: line 6, col 10, disallowSpacesInAnonymousFunctionExpression: Illegal space before opening round brace
* [ ] src/wp-includes/js/wp-media-widget.js: line 6, col 10, disallowSpacesInFunctionExpression: Illegal space before opening round brace
* [ ] src/wp-includes/js/wp-media-widget.js: line 39, col 3, requireVarDeclFirst: Variable declarations must be the first statements of a function scope.
* [ ] src/wp-includes/js/wp-media-widget.js: line 39, col 29, requireSpacesInsideParentheses: Missing space after opening round bracket
* [ ] src/wp-includes/js/wp-media-widget.js: line 39, col 32, requireSpacesInsideParentheses: Missing space before closing round bracket
* [ ] src/wp-includes/js/wp-media-widget.js: line 65, col 38, requireCapitalizedComments: Comments must start with an uppercase letter, unless it is part of a textblock
* [ ] src/wp-includes/js/wp-media-widget.js: line 95, col 59, requirePaddingNewLinesBeforeLineComments: Line comments must be preceded with a blank line
* [ ] src/wp-includes/js/wp-media-widget.js: line 116, col 68, requirePaddingNewLinesBeforeLineComments: Line comments must be preceded with a blank line
* [ ] src/wp-includes/js/wp-media-widget.js: line 128, col 49, disallowSpacesInFunctionExpression: Illegal space before opening round brace
* [ ] src/wp-includes/js/wp-media-widget.js: line 128, col 49, disallowSpacesInAnonymousFunctionExpression: Illegal space before opening round brace
* [ ] src/wp-includes/js/wp-media-widget.js: line 150, col 41, requireSpacesInsideParentheses: Missing space after opening round bracket
* [ ] src/wp-includes/js/wp-media-widget.js: line 150, col 42, requireSpacesInsideParentheses: Missing space before closing round bracket
* [ ] src/wp-includes/js/wp-media-widget.js: line 150, col 71, requireSpacesInsideParentheses: Missing space after opening round bracket
* [ ] src/wp-includes/js/wp-media-widget.js: line 150, col 72, requireSpacesInsideParentheses: Missing space before closing round bracket

## PHPCS

* [ ] src/wp-includes/script-loader.php:765:9: error - Inline comments must end in full-stops, exclamation marks, or question marks (Squiz.Commenting.InlineComment.InvalidEndChar)
* [ ] src/wp-includes/script-loader.php:770:30: error - Each line in an array declaration must end in a comma (WordPress.Arrays.ArrayDeclaration.NoComma)
* [ ] src/wp-includes/script-loader.php:876:43: error - String "/wp-includes/css/wp-media-widget.css" does not require double quotes; use single quotes instead (Squiz.Strings.DoubleQuoteUsage.NotRequired)
* [ ] src/wp-includes/widgets.php:1469:5: error - Expected 1 spaces after opening bracket; 0 found (PEAR.Functions.FunctionCallSignature.SpaceAfterOpenBracket)
* [ ] src/wp-includes/widgets.php:1469:5: error - Expected 1 spaces before closing bracket; 0 found (PEAR.Functions.FunctionCallSignature.SpaceBeforeCloseBracket)
* [ ] src/wp-includes/widgets/class-wp-widget-media.php:1:1: error - Class file names should be based on the class name with "class-" prepended. Expected class-wp-media-widget.php, but found class-wp-widget-media.php. (WordPress.Files.FileName.InvalidClassFileName)
* [ ] src/wp-includes/widgets/class-wp-widget-media.php:18:13: error - Missing member variable doc comment (Squiz.Commenting.VariableComment.Missing)
* [ ] src/wp-includes/widgets/class-wp-widget-media.php:98:14: error - Expected next thing to be an escaping function (see Codex for 'Data Validation'), not '$output' (WordPress.XSS.EscapeOutput.OutputNotEscaped)
* [ ] src/wp-includes/widgets/class-wp-widget-media.php:116:9: error - Inline comments must end in full-stops, exclamation marks, or question marks (Squiz.Commenting.InlineComment.InvalidEndChar)
* [ ] src/wp-includes/widgets/class-wp-widget-media.php:151:9: error - Inline comments must end in full-stops, exclamation marks, or question marks (Squiz.Commenting.InlineComment.InvalidEndChar)
* [ ] src/wp-includes/widgets/class-wp-widget-media.php:210:26: error - Expected 1 spaces before closing bracket; 0 found (PEAR.Functions.FunctionCallSignature.SpaceBeforeCloseBracket)
* [ ] src/wp-includes/widgets/class-wp-widget-media.php:212:26: error - Each line in an array declaration must end in a comma (WordPress.Arrays.ArrayDeclaration.NoComma)
* [ ] src/wp-includes/widgets/class-wp-widget-media.php:231:13: warning - The method parameter $widget_id is never used (Generic.CodeAnalysis.UnusedFunctionParameter.Found)
* [ ] src/wp-includes/widgets/class-wp-widget-media.php:232:23: warning - Not using strict comparison for in_array; supply true for third argument. (WordPress.PHP.StrictInArray.MissingTrueStrict)
* [ ] src/wp-includes/widgets/class-wp-widget-media.php:237:22: error - Each line in an array declaration must end in a comma (WordPress.Arrays.ArrayDeclaration.NoComma)
* [ ] src/wp-includes/widgets/class-wp-widget-media.php:252:13: warning - The method parameter $widget_id is never used (Generic.CodeAnalysis.UnusedFunctionParameter.Found)
* [ ] src/wp-includes/widgets/class-wp-widget-media.php:253:23: warning - Not using strict comparison for in_array; supply true for third argument. (WordPress.PHP.StrictInArray.MissingTrueStrict)
* [ ] src/wp-includes/widgets/class-wp-widget-media.php:258:22: error - Each line in an array declaration must end in a comma (WordPress.Arrays.ArrayDeclaration.NoComma)
* [ ] src/wp-includes/widgets/class-wp-widget-media.php:273:13: warning - The method parameter $instance is never used (Generic.CodeAnalysis.UnusedFunctionParameter.Found)
* [ ] src/wp-includes/widgets/class-wp-widget-media.php:309:36: error - Concat operator must be surrounded by a single space (Squiz.Strings.ConcatenationSpacing.PaddingFound)
* [ ] src/wp-includes/widgets/class-wp-widget-media.php:319:8: error - Function return type is not void, but function has no return statement (Squiz.Commenting.FunctionComment.InvalidNoReturn)
* [ ] src/wp-includes/widgets/class-wp-widget-media.php:337:40: error - Expected next thing to be an escaping function (see Codex for 'Data Validation'), not '$this' (WordPress.XSS.EscapeOutput.OutputNotEscaped)
* [ ] src/wp-includes/widgets/class-wp-widget-media.php:338:55: error - Expected next thing to be an escaping function (see Codex for 'Data Validation'), not '$this' (WordPress.XSS.EscapeOutput.OutputNotEscaped)
* [ ] src/wp-includes/widgets/class-wp-widget-media.php:338:108: error - Expected next thing to be an escaping function (see Codex for 'Data Validation'), not '$this' (WordPress.XSS.EscapeOutput.OutputNotEscaped)
* [ ] src/wp-includes/widgets/class-wp-widget-media.php:345:68: error - Expected next thing to be an escaping function (see Codex for 'Data Validation'), not '$widget_id' (WordPress.XSS.EscapeOutput.OutputNotEscaped)
* [ ] src/wp-includes/widgets/class-wp-widget-media.php:347:17: error - Line indented incorrectly; expected 3 tabs, found 4 (Generic.WhiteSpace.ScopeIndent.IncorrectExact)
* [ ] src/wp-includes/widgets/class-wp-widget-media.php:348:26: error - Expected next thing to be an escaping function (see Codex for 'Data Validation'), not '$this' (WordPress.XSS.EscapeOutput.OutputNotEscaped)
* [ ] src/wp-includes/widgets/class-wp-widget-media.php:348:60: error - Expected next thing to be an escaping function (see Codex for 'Data Validation'), not '$widget_id' (WordPress.XSS.EscapeOutput.OutputNotEscaped)
* [ ] src/wp-includes/widgets/class-wp-widget-media.php:348:72: error - Expected next thing to be an escaping function (see Codex for 'Data Validation'), not '$instance' (WordPress.XSS.EscapeOutput.OutputNotEscaped)
* [ ] src/wp-includes/widgets/class-wp-widget-media.php:349:26: error - Expected next thing to be an escaping function (see Codex for 'Data Validation'), not '$this' (WordPress.XSS.EscapeOutput.OutputNotEscaped)
* [ ] src/wp-includes/widgets/class-wp-widget-media.php:349:68: error - Expected next thing to be an escaping function (see Codex for 'Data Validation'), not '$widget_id' (WordPress.XSS.EscapeOutput.OutputNotEscaped)
* [ ] src/wp-includes/widgets/class-wp-widget-media.php:349:80: error - Expected next thing to be an escaping function (see Codex for 'Data Validation'), not '$instance' (WordPress.XSS.EscapeOutput.OutputNotEscaped)
* [ ] src/wp-includes/widgets/class-wp-widget-media.php:350:17: error - Line indented incorrectly; expected 3 tabs, found 4 (Generic.WhiteSpace.ScopeIndent.IncorrectExact)
* [ ] src/wp-includes/widgets/class-wp-widget-media.php:436:25: warning - Not using strict comparison for in_array; supply true for third argument. (WordPress.PHP.StrictInArray.MissingTrueStrict)

File: src/wp-includes/class-wp-customize-manager.php [hunk 1]
```diff
@@ -1807,6 +1807,10 @@ public function customize_preview_settings() {
 				'stylesheet' => $this->get_stylesheet(),
 				'active'     => $this->is_theme_active(),
 			),
+			'media' => array(
+				'audioLibrary' => apply_filters( 'wp_audio_shortcode_library', 'mediaelement' ),
```
Comment (westonruter) [inline src/wp-includes/class-wp-customize-manager.php]: These need `/** ... */` filter docs.

File: src/wp-includes/widgets/class-wp-widget-media.php [hunk 2]
```diff
@@ -0,0 +1,454 @@
...
+		$instance = $old_instance;
+
+		// ID and title.
+		$instance['id']    = (int) $new_instance['id'];
+		$instance['title'] = sanitize_text_field( $new_instance['title'] );
+
+		// Everything else.
+		$instance['align'] = sanitize_text_field( $new_instance['align'] );
+		$instance['size']  = sanitize_text_field( $new_instance['size'] );
+		$instance['link']  = sanitize_text_field( $new_instance['link'] );
```
Comment (westonruter) [inline src/wp-includes/widgets/class-wp-widget-media.php]: These sanitizing functions should probably be more specific.

File: src/wp-includes/widgets/class-wp-widget-media.php [hunk 3]
```diff
@@ -0,0 +1,454 @@
...
+	 * @return string
+	 */
+	private function render_image( $attachment, $widget_id, $instance ) {
+		$has_caption   = ( ! empty( $attachment->post_excerpt ) );
+
+		$img_attrs = array(
+			'data-id' => $widget_id,
+			'title'   => $attachment->post_title,
+			'class'   => 'image wp-image-' . $attachment->ID,
+			'style'   => 'width: 100%; height: auto;',
```
Comment (celloexpressions) [inline src/wp-includes/widgets/class-wp-widget-media.php]: Let's not do this. The core-provided CSS should already cover this, and themes should have their own CSS doing the same thing. Inlining it is far from ideal.

File: src/wp-includes/widgets/class-wp-widget-media.php [hunk 4]
```diff
@@ -0,0 +1,454 @@
...
+	 * Get styles for responsifying the widget
+	 *
+	 * @since 4.8.0
+	 * @access private
+	 *
+	 * @param WP_Post $attachment Attachment object.
+	 * @param string  $widget_id  Widget ID.
+	 * @return string styles for responsive media
+	 */
+	private function get_responsive_style( $attachment, $widget_id ) {
```
Comment (celloexpressions) [inline src/wp-includes/widgets/class-wp-widget-media.php]: Again, I think there's a better way to do this. `wp-includes/css/wp-media-widget.css` seems to already be doing this? Shouldn't be necessary to do inline `<style>`s with JS.

File: src/wp-includes/widgets/class-wp-widget-media.php [hunk 5]
```diff
@@ -0,0 +1,454 @@
...
+
+	/**
+	 * Creates and returns a link for an attachment
+	 *
+	 * @param WP_Post $attachment Attachment object.
+	 * @param string  $type       link type.
+	 * @return string
+	 */
+	private function create_link_for( $attachment, $type = '' ) {
+		$url = '#';
```
Comment (celloexpressions) [inline src/wp-includes/widgets/class-wp-widget-media.php]: This function should probably default to providing no link, not linking to `#`.

**Comment by westonruter** [comment]
Closing per https://core.trac.wordpress.org/ticket/32417#comment:138

New development will be done on https://github.com/xwp/wp-core-media-widgets

**Pull request closed without merging**